vcsutils/utils.go

```diff
@@ -132,8 +133,113 @@ func safeCopy(targetFile *os.File, reader *tar.Reader) error {
 // DiscardResponseBody prepare http response body for closing
 func DiscardResponseBody(resp *http.Response) error {
 	if resp != nil {
-		_, err := io.Copy(ioutil.Discard, resp.Body)
+		_, err := io.Copy(io.Discard, resp.Body)
 		return err
 	}
 	return nil
 }
+
+// GetZeroValue returns the zero value of type T
+func GetZeroValue[T any]() T {
+	return *new(T)
+}
+
+// DefaultIfNotNil checks:
+// 1. If the pointer is nil, return the zero value of the type
+// 2. If the pointer isn't nil, return the value of the pointer.
+func DefaultIfNotNil[T any](val *T) T {
+	if val == nil {
+		return GetZeroValue[T]()
+	}
+	return *val
+}
+
+func AddBranchPrefix(branch string) string {
+	if !strings.HasPrefix(branch, branchPrefix) {
+		branch = fmt.Sprintf("%s%s", branchPrefix, branch)
+	}
+	return branch
+}
+
+// Unzip a file in src to dest path
+func Unzip(source, destination string) (err error) {
+	// Open the zip file
+	reader, err := zip.OpenReader(source)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		e := reader.Close()
+		if err == nil {
+			err = e
+		}
+	}()
+
+	// Get the absolute destination path
+	destination, err = filepath.Abs(destination)
+	if err != nil {
+		return err
+	}
+
+	// Iterate over zip files inside the archive and unzip each of them
+	for _, f := range reader.File {
+		err := unzipFile(f, destination)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func unzipFile(f *zip.File, destination string) error {
+	fullFilePath := filepath.Join(destination, filepath.Clean(f.Name))
+	// Check if file paths are not vulnerable to Zip Slip
+	if !strings.HasPrefix(fullFilePath, filepath.Clean(destination)+string(os.PathSeparator)) {
+		return fmt.Errorf("invalid file path: %s", fullFilePath)
+	}
+
+	// Create directory tree
+	if f.FileInfo().IsDir() {
+		if err := os.MkdirAll(fullFilePath, os.ModePerm); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if err := os.MkdirAll(filepath.Dir(fullFilePath), os.ModePerm); err != nil {
+		return err
+	}
+
+	// Create a destination file for unzipped content
+	destinationFile, err := os.OpenFile(filepath.Clean(fullFilePath), os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := destinationFile.Close(); err != nil {
+			return
+		}
```

Comment:
This defer (and the other one inside this function) should return the errorm if there isn't already an error return by this function.

Reply:
done